### PR TITLE
test: run mempool_compatibility.py even with wallet disabled

### PR DESCRIPTION
**Pull request by mjdietzx**
Another functional test rewritten as proposed in https://github.com/bitcoin/bitcoin/issues/20078

### CODE REVIEW

File: test/functional/mempool_compatibility.py [hunk 1]
```diff
@@ -33,18 +33,24 @@ def setup_network(self):
...
+        old_wallet.generate(1)
+        new_wallet.generate(1)
+        old_node.generate(100)
+        self.connect_nodes(0, 1)
+        self.sync_all()
 
+        self.stop_node(1)
         self.log.info("Add a transaction to mempool on old node and shutdown")
-        old_tx_hash = old_node.sendtoaddress(recipient, 0.0001)
+        old_tx_hash = old_wallet.send_self_transfer(from_node=old_node)['txid']
```
Comment (MarcoFalke) [inline test/functional/mempool_compatibility.py:53]: no need to replace this. old nodes always have a wallet

File: test/functional/mempool_compatibility.py [hunk 2]
```diff
@@ -33,18 +33,24 @@ def setup_network(self):
...
 
-        old_node = self.nodes[0]
-        new_node = self.nodes[1]
-        recipient = old_node.getnewaddress()
-        self.stop_node(1)
+        old_node, new_node = self.nodes
+        old_wallet = MiniWallet(old_node)
+        new_wallet = MiniWallet(new_node)
+
+        # Add enough mature utxos to the wallets so that all txs spend confirmed coins.
```
Comment (MarcoFalke) [inline test/functional/mempool_compatibility.py:44]: No need for this comment? There is only one inputs, so this is clear from reading the code.

**Review by maflcko — COMMENTED** [review]
Concept ACK

**Comment by mjdietzx** [comment]
The 1st commit https://github.com/bitcoin/bitcoin/pull/20688/commits/f2b7cb56c8cbec9c6a2a8ec88143756974f6860f is ready to go now

I realized that I could extend the test coverage a bit, see 2nd commit https://github.com/bitcoin/bitcoin/pull/20688/commits/6a1e9adeaea8a12a23ec4be5d59cec4fccd0c196. I know it's not totally related to this PR. Maybe it seems more appropriate for a test like this to go in `mempool_accept.py`, but this test-case does fit in naturally to `MempoolCompatibilityTest`, and seems a little tricky to do in `mempool_accept.py` the way that test is setup.

What do you think is the best way forward? If you want it in this PR I can squash the commits, otherwise I'll ditch the 2nd commit

File: test/functional/mempool_compatibility.py [hunk 3]
```diff
@@ -69,7 +68,23 @@ def run_test(self):
         self.stop_node(1)
 
         self.log.info("Move mempool.dat from new to old node")
-        os.rename(new_node_mempool, old_node_mempool)
+        shutil.copyfile(new_node_mempool, old_node_mempool)
+
+        # Coinbase is only valid in a block, not as a loose transaction.
+        # Because the nodes have not been synced, old_node will not accept unbroadcasted tx to its mempool.
```
Comment (instagibbs) [inline test/functional/mempool_compatibility.py:74]: can you convert this "nodes have not been synced" statement into an assertion? May catch future test regressions, and is more self-explanatory

File: test/functional/mempool_compatibility.py [hunk 4]
```diff
@@ -69,7 +68,23 @@ def run_test(self):
         self.stop_node(1)
 
         self.log.info("Move mempool.dat from new to old node")
-        os.rename(new_node_mempool, old_node_mempool)
+        shutil.copyfile(new_node_mempool, old_node_mempool)
+
+        # Coinbase is only valid in a block, not as a loose transaction.
```
Comment (instagibbs) [inline test/functional/mempool_compatibility.py:73]: I don't understand why this comment is here

**Comment by instagibbs** [comment]
concept ACK

**Comment by mjdietzx** [comment]
Hey @instagibbs thanks for the review. How does this look? https://github.com/bitcoin/bitcoin/pull/20688/commits/174da9f63a51747e7297949414edba6bf9d673dc.  lmk if you have any ideas on how this can be done better - I know it doesn't perfectly address your comments, but seemed like a good balance to me without adding too much code or extra stuff

**Comment by instagibbs** [comment]
better thanks, you can squash the last two commits together :+1: 

ACK otherwise

**Review by jonatack — COMMENTED** [review]
ACK bff5aca27d0d66587c0155642a9c6cd88228ed29 code review, tested with bitcoind built with the wallet and without

File: test/functional/mempool_compatibility.py [hunk 5]
```diff
@@ -33,13 +38,15 @@ def setup_network(self):
             None,
         ])
         self.start_nodes()
-        self.import_deterministic_coinbase_privkeys()
+        self.init_wallet(0)  # only import_deterministic_coinbase_privkeys for old_node
```
Comment (MarcoFalke) [inline test/functional/mempool_compatibility.py:41]: Does this line need to change? If yes, why?

Comment (mjdietzx) [reply test/functional/mempool_compatibility.py]: Good point - this change is not necessary. I got rid of it and force pushed

File: test/functional/mempool_compatibility.py [hunk 6]
```diff
@@ -48,28 +55,44 @@ def run_test(self):
...
+        shutil.copyfile(new_node_mempool, old_node_mempool)
 
-        self.log.info("Start old node again and verify mempool contains both txs")
+        self.log.info("Start old node again and verify mempool does not contain unbroadcasted_tx_hash")
         self.start_node(0, ['-nowallet'])
         assert old_tx_hash in old_node.getrawmempool()
+        # Because the nodes haven't been synced, old_node doesn't have the block that contains the coinbase that unbroadcasted_tx spends.
+        # As coinbases are only valid in a block, and not as loose txns, unbroadcasted_tx won't pass `MemPoolAccept::PreChecks`.
+        assert_raises_rpc_error(-5, "Block not found", old_node.getblock, blockhash=block_hash)
+        assert unbroadcasted_tx_hash not in old_node.getrawmempool()
```
Comment (MarcoFalke) [inline test/functional/mempool_compatibility.py:83]: I don't understand why this test is being changed. The only change should be replacing the bitcoin core wallet with miniwallet. If you want to do additional changes, they can be done in a separate commit or pull request.

File: test/functional/mempool_compatibility.py [hunk 7]
```diff
@@ -57,14 +59,17 @@ def run_test(self):
...
         assert unbroadcasted_tx_hash in new_node.getrawmempool()
         mempool = new_node.getrawmempool(True)
         assert mempool[unbroadcasted_tx_hash]['unbroadcast']
-        self.stop_node(1)
 
-        self.log.info("Move mempool.dat from new to old node")
+        self.log.info("Sync the nodes and move mempool.dat from new to old node and shutdown")
+        self.start_node(0, ['-nowallet'])
+        self.connect_nodes(0, 1)
+        self.sync_blocks()
```
Comment (MarcoFalke) [inline test/functional/mempool_compatibility.py:71]: Would be nice to move up the sync where the blocks are generated to not accidentally change what the test is testing and preserving previous test behavior

Comment (mjdietzx) [reply test/functional/mempool_compatibility.py]: Good point, that will also reduce the diff. Done

**Comment by MarcoFalke** [comment]
review ACK a7599c80ebb9579df45e2d6ccf3168302cf42f03 didn't test

**Pull request merged**